hypervisor/src/kvm/mod.rs

```diff
@@ -696,9 +696,35 @@ impl cpu::Vcpu for KvmVcpu {
         match self.fd.run() {
             Ok(run) => match run {
                 #[cfg(target_arch = "x86_64")]
-                VcpuExit::IoIn(addr, data) => Ok(cpu::VmExit::IoIn(addr, data)),
+                VcpuExit::IoIn(addr, data) => {
+                    if let Some(ref vmmops) = self.vmmops {
+                        let ret = vmmops.lock().unwrap().pio_read(addr.into(), data);
+                        if ret.is_ok() {
+                            return Ok(cpu::VmExit::Ignore);
+                        }
+                    }
+                    Err(cpu::HypervisorCpuError::RunVcpu(anyhow!(
```

Comment:
I think we want the following logic instead:

```Rust
                VcpuExit::IoIn(addr, data) => {
                    if let Some(ref vmmops) = self.vmmops {
                        return vmmops
                            .lock()
                            .unwrap()
                            .pio_read(addr.into(), data)
                            .map(|_| cpu::VmExit::Ignore)
                            .map_err(|e| cpu::HypervisorCpuError::RunVcpu(e.into()));
                    }

                    Ok(cpu::VmExit::IoIn(addr, data))
                }
```

1. If we have a `VmmOps`, we handle the VM exit through it and return the ops result itself.
1. If we don't have a `VmmOps`, we offload the VM exit back to the VMM.

We could argue about the second step and make it return an error when it's missing a `VmmOps`, but since we don't have a way to make it mandatory for now, we should gracefully handle this option.

On Cloud Hypervisor, we would error when receiving a `VmExit::IoIn` from the hypervisor crate.

Reply:
make sense Samuel, will adopt this logic